### PR TITLE
Add `size_of` and `size_of_val` and `align_of` and `align_of_val` to the prelude

### PR DESCRIPTION
**Pull request by joshtriplett**
(Note: need to update the PR to add `align_of` and `align_of_val`, and remove the second commit with the myriad changes to appease the lint.)

Many, many projects use `size_of` to get the size of a type. However,
it's also often equally easy to hardcode a size (e.g. `8` instead of
`size_of::<u64>()`). Minimizing friction in the use of `size_of` helps
ensure that people use it and make code more self-documenting.

The name `size_of` is unambiguous: the name alone, without any prefix or
path, is self-explanatory and unmistakeable for any other functionality.
Adding it to the prelude cannot produce any name conflicts, as any local
definition will silently shadow the one from the prelude. Thus, we don't
need to wait for a new edition prelude to add it.

### CODE REVIEW

**Comment by taiki-e** [comment]
> Thus, we don't need to wait for a new edition prelude to add it.


How does this interact with unused_imports lint (https://github.com/rust-lang/rust/issues/121708)?

If the latest nightly with this patch shows a warning but stable does not have these functions in prelude, that seems to be an unfortunate situation for many users who test both nightly and stable in CI.

**Comment by joshtriplett** [comment]
@taiki-e I have the impression that in the past we didn't warn about redundant imports of things in the prelude. Let's see what policy comes out of #121708.

**Comment by cuviper** [comment]
r? libs-api

File: src/tools/rust-analyzer/crates/hir-ty/src/display.rs [hunk 1]
```diff
@@ -2,10 +2,7 @@
 //! HIR back into source code, and just displaying them for debugging/testing
 //! purposes.
 
-use std::{
-    fmt::{self, Debug},
-    mem::size_of,
-};
```
Comment (Veykril) [inline src/tools/rust-analyzer/crates/hir-ty/src/display.rs:8]: r-a builds against stable out of tree so this will fail to build when synced back

Comment (joshtriplett) [reply src/tools/rust-analyzer/crates/hir-ty/src/display.rs:8]: @Veykril I'll switch it to a temporary `allow` instead.

**Comment by RalfJung** [comment]
> Can we just have u64::BYTES? 

Or `u64::SIZE`? `size_of` also doesn't say "bytes".

**Comment by joshtriplett** [comment]
> Could we then give a way that's less awkward?

I mean, I'd love to have a less awkward syntax than the turbofish in the *general* case, but we don't currently have one. I don't think there's a good special-case syntax for this, unless we wanted to provide a macro (e.g. `size_of!(T)`).

> Can we just have `u64::BYTES`?

I'd love to see that, but that doesn't help for other types, including user-defined types.

> For example, why not `align_of` too?

Zero objections, but I was trying to make a smaller change that would be less likely to have trouble getting approved.

> Procedurally, BTW, why is this a PR?

One way or another this needs rationale and an FCP. In this case, I put the rationale in the PR and commit message. I have no problem writing an RFC, but I don't know if the rationale needs substantial expansion or if we'd have consensus already with the current rationale.

Also, depending on the outcome of https://github.com/rust-lang/rust/issues/121708 we may want this to be edition-specific after all. Either outcome would be fine with me.

**Comment by oli-obk** [comment]
> I'd love to see that, but that doesn't help for other types, including user-defined types

Could be an assoc const on `Sized`

**Comment by workingjubilee** [comment]
I think that if this had been offered _before_ the `redundant_imports` lint, it would've been okay to use a more lightweight procedure.

With, I think we actually do need to sit down and wrack our brains for whether it's justified to make all -Dwarnings codebases rip this out everywhere.

**Comment by jhpratt** [comment]
> Could be an assoc const on `Sized`

Something trivial like `const SIZE: usize = core::mem::size_of::<Self>();`, I presume? My immediate thought is how this works given that `Sized` is a marker trait, which afaik requires the trait to contain nothing. Although I suppose it is already a lang item, so we could do whatever we wanted?

File: src/tools/rust-analyzer/crates/hir-ty/src/display.rs (hunk 1 above)
Comment (jhpratt) [reply src/tools/rust-analyzer/crates/hir-ty/src/display.rs:8]: Does the same reasoning apply to clippy and miri? I notice that files in those directories are all `allow`ed.

**Comment by cuviper** [comment]
> Something trivial like `const SIZE: usize = core::mem::size_of::<Self>();`, I presume? My immediate thought is how this works given that `Sized` is a marker trait, which afaik requires the trait to contain nothing.

I'm not sure about the exact rules, but the compiler does accept this addition right now. I think that restriction is about `auto trait`, while `Sized` is more deeply built-in (`#[rustc_deny_explicit_impl]`).

**Comment by programmerjake** [comment]
> Could be an assoc const on `Sized`

My concern with that is that it could easily conflict with associated constants on tons of other traits, I expect `SIZE` to be a relatively common associated constant name.

e.g. [this search](https://github.com/search?q=lang%3ARust+trait+%2F%5E%5B+%5Ct%5D%2Bconst+SIZE%3A%2F&type=code) has 1.9k results and I count several `SIZE` associated constants in common crates on just the first page.

**Comment by saethlin** [comment]
Yeah, we had a lot of conflicts when `BITS` was added to the built-in numeric types.

**Comment by cuviper** [comment]
`const SIZE_OF` is much less common, but it's a bit Yoda-like to say `T::SIZE_OF`.

File: src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #![allow(clippy::redundant_clone, clippy::unnecessary_operation, clippy::incompatible_msrv)]
 #![warn(clippy::manual_non_exhaustive, clippy::borrow_as_ptr, clippy::manual_bits)]
 
+#[allow(unused_imports)]
 use std::mem::{size_of, size_of_val};
```
[the remark this answers is not in the document]
Comment (joshtriplett) [reply src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs:5]: @flip1995 This is in clippy, and clippy is maintained in a separate tree, which doesn't necessarily move in lockstep with rust nightly. Using an `allow` allows this to build on both stable and nightly.

**Comment by joshtriplett** [comment]
I think it's going to end up being important to settle the question of the redundant imports lint before we decide on 1) whether to do this and 2) what approach to use to do this (e.g. edition vs everywhere).

File: src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs:5]: > which doesn't necessarily move in lockstep with rust nightly

It has been in lockstep for years now

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting.

Everyone was in favor of adding `align_of` in addition to `size_of`. (I've retitled accordingly, though I haven't updated the PR yet.)

We want to wait for the conclusion on the lint change before actually merging this. The amount of warnings that that lint change produces for adding a function to the prelude is very high, making it more difficult to introduce functions in the prelude.

The lint is very likely to impact libs-api policy on additions to the prelude. If the lint stays as-is, we're likely to only add items to the prelude in new editions. If the lint is adjusted somehow (e.g. by reverting it or by adding some kind of exception for preludes or re-exports), then we can add things to the common prelude for all editions.

Meanwhile, though, there seemed to be consensus to add these to the prelude, conditional on resolving the lint issue (and then deciding whether to use the 2024 prelude or the common prelude). So:

@rfcbot merge
@rfcbot concern settle-redundant-import-lint

**Comment by joshtriplett** [comment]
Pushed the simple version that *just* adds this to the prelude. Depending on the decision in https://github.com/rust-lang/rust/issues/121708 , we can either use this version, or a version that only adds these to the 2024 prelude.

File: src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (flip1995) [reply src/tools/clippy/tests/ui-toml/min_rust_version/min_rust_version.rs]: Once this is pulled into Clippy, the nightly will be updated to a nightly that already has those changes done in `rustc`.

**Comment by tgross35** [comment]
> > Could we then give a way that's less awkward?
> 
> I mean, I'd love to have a less awkward syntax than the turbofish in the _general_ case, but we don't currently have one. I don't think there's a good special-case syntax for this, unless we wanted to provide a macro (e.g. `size_of!(T)`).

Driveby comment: I really wouldn't mind having `size_of!` and `align_of!` that work for both types and expressions, and return an untyped integer literal. Having two different functions and needing turbofish for one is pretty awkward compared to other languages (C's `sizeof`/`_Alignof` and Zig's `@sizeOf`/`@alignOf` seem quite elegant by comparison).

Even if that doesn't happen, having something in the prelude here will be quite nice.

**Comment by jhpratt** [comment]
@tgross35 How would that work? A type and value could have the same name, and the value isn't even necessarily of that type.

**Comment by programmerjake** [comment]
> @tgross35 How would that work? A type and value could have the same name, and the value isn't even necessarily of that type.

that kind of thing is commonly used by [`syn::custom_keyword!`](https://docs.rs/syn/latest/syn/macro.custom_keyword.html), where it makes both a struct and a function with the same name.

**Comment by jhpratt** [comment]
That still wouldn't work, as `const`s and `static`s could share the name and be imported at the same time as a type. Don't get me wrong — I think it would be great. But I don't see how it _could_ work.

Anyway this is more of an aside to the PR, so let's take this to Zulip if you'd prefer?

**Comment by joshtriplett** [comment]
(This needs to wait as beta hasn't been updated yet.)

**Comment by BurntSushi** [comment]
Sorry for being late to the party.

While I don't think I feel strongly enough to register a concern, these functions don't seem like they belong in the prelude to me. There are assuredly contexts in which I use `size_of` quite a lot, but they tend to be specific contexts or related to lower level optimizations. (I'm being hand wavy.) But it doesn't have the feel, to me, of something that is universally applicable. `align_of` is something that I use way less often than `size_of`, and so I feel that the argument for including it in the prelude is even weaker. I suppose it (along with `size_of_val` and `align_of_val`) is being included here to complete the picture provided by `size_of`. Which I guess makes sense, but still, these just don't seem widely applicable enough to belong in the prelude.

If I look in the prelude for other things that maybe shouldn't be there, the only thing that really pops out at me (with the exception of some macros) is `drop`. It feels very similar to `size_of` where it isn't really universally applicable enough to have a spot in the prelude.

**Comment by joshtriplett** [comment]
@BurntSushi I feel like the implicit criteria you're using for the prelude is different than the criteria I think we should be using.

Adding items (other than traits) to the prelude will never produce conflicts with existing code, since we allow shadowing. So, the main question I think we need to ask is "will anyone be confused if they see this item unqualified, without seeing anything introduce it first?". A particular flavor of that confusion is "will anyone be confused and think this is likely to be something local to the program or one of its dependencies?".

I also think the value of "seeing it introduced first" is limited, insofar as that *still* typically requires people to scroll up to the top of the module; *any* unqualified import should generally have a clear meaning to people reading, say, a modified line of code in a patch with just 3 lines of context on either side, without needing to look at the top of the module or get IDE assistance.

I think `size_of` very clearly meets that criteria. People aren't going to mistake it for anything else, and other crates are exceptionally unlikely to define a conflicting `size_of` that people are expected to use unqualified. Even unqualified, `size_of` has one clear obvious unambiguous meaning.

Beyond that, it doesn't need to be used in the *majority* of programs. Many items in the prelude aren't used in the *majority* of programs. I'd venture a guess that *most* programs don't use `Unpin`, or `ExactSizeIterator`. I would *bet* that more programs use `size_of` than use `ExactSizeIterator`. But there's still value in having all of those in the prelude.

On top of that, I think some people will feel as I do: "if `size_of` is no trouble at all to use, I'll write `size_of::<u64>()` instead of `8`". This is compounded by other languages having the equivalent operation built in.

**Comment by joshtriplett** [comment]
@ehuss False alarm; a rebase caused the `src/tools/cargo` submodule to end up in the commit. Fixed now. (This seems to happen quite often; rebase and submodules don't seem to play nice.)

**Comment by BurntSushi** [comment]
@joshtriplett Yeah I agree that `size_of` meets the criteria you outlined. But I feel like those criteria are coming at this from a "what are the downsides of something being in the prelude" versus where I'm coming from which is more like "I don't think `size_of` is used enough to deserve a spot in the prelude in the first place." A prelude forms a fundamental vocabulary, and increasing that has costs. It increases the surface area that is in scope in almost every Rust program. Each individual addition is arguably marginal, but if we don't stop to consider the overall effect of adding too much, we could easily wind up in a situation where our prelude becomes too big.

Like, I'm not totally opposed to this. I get it. I use `size_of` a lot and I get the reasoning that making it easier to use _might_ decrease the incidence of less robust code. I even agree that it might indeed do that. I'm just not sure it's worth the expansion of a prelude which really ought to be small. Every time we add to it, we use up a little more of our budget. Of course, we don't have any concrete notion of what our budget even is, which is what makes this sort of thing very hand wavy.

**Comment by joshtriplett** [comment]
@BurntSushi
> what are the downsides of something being in the prelude

This *is* the angle I'm coming from, but I do want to take into account the criteria you're describing. It sounds like you're saying that items in the prelude have a weight in aggregate, such that even if one individual item meets all our other criteria, doubling the size of the prelude might be a cost in itself.

I'm not sure if I'd entirely agree with that criteria, insofar as I don't think that everyone needs to know about every item in the prelude. For instance, I think `ExactSizeIterator`, while self-explanatory in *semantic*, is probably something people would need to look up the exact definition of before using it, unless they'd worked with it before. Nonetheless, I don't think it does any *harm* to anyone to have it in the prelude; anyone not familiar with it will look it up. I think the *harm* it would do is if we polluted the namespace; for instance, we should never add a function `new` to the prelude because it'd be incredibly confusing.

That aside, though, even if we *did* agree upon that criteria, and agreed on the incremental cost of an addition to the prelude, I think `size_of` clears the bar for inclusion.

**Comment by BurntSushi** [comment]
> I'm not sure if I'd entirely agree with that criteria, insofar as I don't think that everyone needs to know about every item in the prelude. For instance, I think `ExactSizeIterator`, while self-explanatory in _semantic_, is probably something people would need to look up the exact definition of before using it, unless they'd worked with it before. Nonetheless, I don't think it does any _harm_ to anyone to have it in the prelude; anyone not familiar with it will look it up, just as they would if it were an imported name. I think the _harm_ it would do is if we polluted the namespace; for instance, we should never add a function `new` to the prelude because it'd be incredibly confusing.

Yeah I mean IDK if I would lobby hard to have `ExactSizeIterator` included in the prelude if we could go back and change things. Like, `Iterator` is clearly warranted but `ExactSizeIterator` isn't (IMO).

But yes, I do think that a large prelude---even if every item is unambiguous and not confusable---is a cost unto itself. It adds more implicit names into every namespace. Like, if I'm reading code and I come across something that I don't know, then yeah, I need to look it up. If I'm in my text editor, rust-analyzer can just take me to the definition. But if I'm reading code outside of my text editor (particularly during code review, but there are several other contexts in which this occurs too), then I usually need to resolve the symbol manually. The first step there is to go look at the imports at the top of the file. If the symbol isn't there, I usually get confused and start hunting around for glob imports.

I think that same sort of effect can happen with a `std` prelude that is "too large." It's not even about whether or not something like `ExactSizeIterator` (or whatever) needs to be looked up, but about where it is even coming from in the first place. If the prelude is reasonably small, then it's reasonable to expect most programmers to keep most of it in their head. And I don't mean a full understanding of every symbol, but just a recognition of, "oh yes, that comes from the prelude, I know how to look it up." A larger prelude, IMO, impacts this in a negative way. It's implicit context that is injected to almost every namespace in Rust.

The problem is that each individual addition is a marginal increase in cost. So in my view, adding something to the prelude isn't just about whether there are downsides limited to that specific addition, but whether its addition justifies that marginal use of our prelude budget in the first place. At least, that's my perspective.

Another way of looking at this is that `size_of` _may_ be deserving of a spot in the prelude, but I'm not convinced any of the rest are. Like I know `size_of` gets used a lot, but I don't think `size_of_val`, `align_of` or `align_of_val` get used even within the same order of magnitude frequency as `size_of` does. But it's hard to quantify precisely.

**Comment by Amanieu** [comment]
@BurntSushi The FCP has completed, but I would like to make sure to check whether your concern over this API is strong enough to block it, or if you are OK with merging this.

**Comment by BurntSushi** [comment]
Thanks for checking! I think given that everyone else on libs-api is a +1 on this, I don't think my concern rises to the level of blocking this. I do remain very skeptical of expanding the prelude without strong justification though.

**Comment by BurntSushi** [comment]
So yes, merge away!

**Comment by traviscross** [comment]
For posterity, here's where we decided (in #123813) to not warn for imports made redundant due to a glob import, which is what will allow for this (and other) expansions of the prelude across all editions.

https://github.com/rust-lang/rust/pull/123813#issuecomment-2089264252

**Comment by liigo** [comment]
> Many, many projects use size_of

I don't think this is the sufficient condition to add `size_of` to the prelude.
Do not many many projects use `std::io::Read` and `std::collections::HashMap`??

**Pull request merged**